docs/Changelog.md

```diff
@@ -14426,35 +14422,9 @@ This version of the operator has been available since version 1 of the 'ai.onnx.
   
   where Loss is a dummy node which computes the minimized objective function.
   
-  Because "W" is an initializer in the called graph, the user can omit it
-  in the input list of the GraphCall and generate the following training
-  graph.
-  
-  ```
-  .---+---- W (declared in the inference graph's initializer list)
-  |   |
-  |   |
-  |   | .-- X_1 --> GraphCall(graph_name="MyInferenceGraph",
-  |   | |            |  |      input_names=["X", "Z"],
-  |   | |            |  |      output_names=["Y"])
-  |   | |   Z_1 -----'  |
-  |   | |    |          V
-  |   | |    |         Y_1 ---> Loss ---> O
-  |   | |    |                    ^
-  |   | |    |                    |
-  |   | `--. |                    L
-  |   |    | |                    |
-  |   |    | |   .----------------'
-  |   |    | |   |
-  |   |    v v   v
-  |   `--> Gradient(xs=["W"], zs=["X_1", "Z_1"], y="O")
-  |        |
-  |        v
-  |      dO_dW (gradient of W)
-  |        |
-  |        v
-  `-----> Sub ----> W_new
-  ```
+  The variable "W" is an optional input in the called graph. If the user omit it,
+  the input names of GraphCall becomes ["X_1", "", "Z_1"]. Notice that
+  this syntax sugar does NOT eliminate the edge connecting "W" and the GraphCall.
```

Comment:
I think "Notice that this syntax does NOT ..." needs to be replaced. I am not sure I understand what exactly it means. I think it is more important to specify the semantics: what happens (at runtime) when the "W" is omitted. I guess this line means that the corresponding initializer's current-value will be passed in implicitly (and it is NOT the case that the initializer value specified in the called-graph will be used). This semantics is, of course, different from the current semantics of a user-call to an inference-graph. This is the reason I preferred to _explicitly_ pass W. But if we allow it to be implicit, we should explain that clearly here.

Reply:
`Notice that this syntax ...` means `W` is passed to the invoked graph by reference. The runtime doesn't make a copy of the current `W`. This also means --- for all invoked graphs without `W` explicitly provided, the same `W` is a global variable and shared (not copied) in every graph.

I agree `explicitly passing W` into the graph makes thing clearer. It implies that all trainable tensors are also inputs of inference graph, which sounds more reasonable.

Reply:
### Pseudo code
My understanding of how training would work (as the spec is currently written) is below (in pseudo python code). I'm trying to emphasise the details that we are discussing, but this should be equivalent to any implementation in terms of results.
```python
def bind(bindings, src, dst):
    for (k,v) in bindings.items():
        dst[k].value = src[v].value
def run(graph, constants, inputs, callable={}):
    """Compute the graph with the given inputs and constants.
    The 'graph' can contain GraphCall to graphs in 'callable'.
    Does NOT mutate graph/constants/inputs."""
    return ...

# Load the model. The model is conceptually trained to some point (0 or more iterations).
model = onnx.load(session.model_path)
# Unlike for inference, we don't allow constant folding of trainable constants
optimize(model, constant_folding_blacklist=trainable_initializers(model))

# Initialize pretty much the same way as for inference.
# I'm explicitly deleting the initializers here to show we are referring only to the
# constant buffers in all graph evaluations henceforth.
# Similarly, these constant buffers could be attached to the graph, I'm making
# them explicit to hopefully make it clearer when they do and don't change.
graph_constants = create_buffers_from(model.graph.initializer)
del model.graph.initializer
algo_constants = [create_buffers_from(info) for info in model.training_info]
for info in model.training_info:
    del info.algorithm.initializer

# Training intialization process (if starting training from scratch);
# overwrites constants.
if session.train_from_scratch:
    for (i, info) in enumerate(model.training_info):
        # Evaluate the initialization graph, which might include generating random
        # tensors. No inputs, and no visibility of initializers outside of itself.
        outputs = run(
            graph = info.initialization,
            constants = create_buffers_from(info.initialization.initializer),
            inputs = [],
            callable = {},
        )
        # Copy the outputs of the initialization graph to the constants referenced
        # via initialization_binding.
        bind(
            bindings = info.initialization_binding,
            src = outputs,
            dst = graph_constants + algo_constants[i],
        )

# Training process
# The spec doesn't specify how data is consumed, but I'll assume we use one batch per
# training info per iteration. Not sure how iterations/training_infos interact.
data = session.get_data_generator()
for iteration in range(session.training_iterations):
    for (i, info) in enumerate(model.training_info):
        # Run the algorithm graph, with access to the inference graph constants
        # and the ability to call the inference graph with a GraphCall node.
        outputs = run(
            graph = info.algorithm,
            constants = graph_constants + algo_constants[i],
            inputs = {**next(data)},
            callable = {
                model.graph.name: (lambda inputs:
                    run(model.graph, graph_constants, inputs, callable={})
                )
            },
        )
        # Copy the outputs of the algorithm graph to the constants referenced
        # via update_binding.
        bind(
            bindings = info.update_bindings,
            src = outputs,
            dst = graph_constants + algo_constants[i],
        )

# Save the model now it has been trained for more iterations
model.graph.initializer = create_initializers_from(graph_constants)
for (i, info) in enumerate(model.training_info):
    info.algorithm.initializer = create_initializers_from(algo_constants[i])
onnx.save(model, session.out_path)
```
Note the constants are updated atomically after each evaluation of a graph -- each evaluation always sees the latest set of constants, but they don't change until after the graph has been evaluated.

As I understand the spec, the result of training should be equivalent to doing this for each iteration:
1. Load model
2. Run algorithm (equivalent to freezing inference graph as an operator and then running algorithm with GraphCall nodes replaced with that operator)
3. Update initializer values from algorithm output based on update_binding
4. Save model

As soon as you've done one iteration the values of the previous iteration are inaccessible (unless you explicitly kept them, or they aren't trainable).

Does this seem right @wschin?

### Regarding optional inputs
It seems the behaviour for optional inputs _within an iteration_ should look identical to the behaviour at inference time. If the input is provided then that tensor name is bound to the input value, if the input is not provided, the tensor name is bound to the constant.

I agree that sentence is confusing:
> Notice that this syntax sugar does NOT eliminate the edge connecting "W" and the GraphCall.

There isn't really an edge between the `W` tensor and the `GraphCall` node when that input isn't provided -- but there is a default constant `W` which is _in_ the graph called by `GraphCall`, and that is what is used for the tensor name `W` within the graph.

### Thoughts
Some thoughts having done the exercise of writing it out:
* Should it be specified how data is consumed? Is the flow is `info[0](data[0]); info[1](data[1]); info[0](data[2]); ...`?
* I'm not sure I understand/remember why having multiple training_infos is helpful, unless I misunderstood how they should be used.
* As I read it, each `algorithm` can only update the inference graph and itself, so each `algorithm` would need to have its own iteration count etc. Is that what you were thinking?

Reply:
> ### Pseudo code
> My understanding of how training would work (as the spec is currently written) is below (in pseudo python code). I'm trying to emphasise the details that we are discussing, but this should be equivalent to any implementation in terms of results.
> 
> ```python
> def bind(bindings, src, dst):
>     for (k,v) in bindings.items():
>         dst[k].value = src[v].value
> def run(graph, constants, inputs, callable={}):
>     """Compute the graph with the given inputs and constants.
>     The 'graph' can contain GraphCall to graphs in 'callable'.
>     Does NOT mutate graph/constants/inputs."""
>     return ...
> 
> # Load the model. The model is conceptually trained to some point (0 or more iterations).
> model = onnx.load(session.model_path)
> # Unlike for inference, we don't allow constant folding of trainable constants
> optimize(model, constant_folding_blacklist=trainable_initializers(model))
> 
> # Initialize pretty much the same way as for inference.
> # I'm explicitly deleting the initializers here to show we are referring only to the
> # constant buffers in all graph evaluations henceforth.
> # Similarly, these constant buffers could be attached to the graph, I'm making
> # them explicit to hopefully make it clearer when they do and don't change.
> graph_constants = create_buffers_from(model.graph.initializer)
> del model.graph.initializer
> algo_constants = [create_buffers_from(info) for info in model.training_info]
> for info in model.training_info:
>     del info.algorithm.initializer
> 
> # Training intialization process (if starting training from scratch);
> # overwrites constants.
> if session.train_from_scratch:
>     for (i, info) in enumerate(model.training_info):
>         # Evaluate the initialization graph, which might include generating random
>         # tensors. No inputs, and no visibility of initializers outside of itself.
>         outputs = run(
>             graph = info.initialization,
>             constants = create_buffers_from(info.initialization.initializer),
>             inputs = [],
>             callable = {},
>         )
>         # Copy the outputs of the initialization graph to the constants referenced
>         # via initialization_binding.
>         bind(
>             bindings = info.initialization_binding,
>             src = outputs,
>             dst = graph_constants + algo_constants[i],
>         )
> 
> # Training process
> # The spec doesn't specify how data is consumed, but I'll assume we use one batch per
> # training info per iteration. Not sure how iterations/training_infos interact.
> data = session.get_data_generator()
> for iteration in range(session.training_iterations):
>     for (i, info) in enumerate(model.training_info):
>         # Run the algorithm graph, with access to the inference graph constants
>         # and the ability to call the inference graph with a GraphCall node.
>         outputs = run(
>             graph = info.algorithm,
>             constants = graph_constants + algo_constants[i],
>             inputs = {**next(data)},
>             callable = {
>                 model.graph.name: (lambda inputs:
>                     run(model.graph, graph_constants, inputs, callable={})
>                 )
>             },
>         )
>         # Copy the outputs of the algorithm graph to the constants referenced
>         # via update_binding.
>         bind(
>             bindings = info.update_bindings,
>             src = outputs,
>             dst = graph_constants + algo_constants[i],
>         )
> 
> # Save the model now it has been trained for more iterations
> model.graph.initializer = create_initializers_from(graph_constants)
> for (i, info) in enumerate(model.training_info):
>     info.algorithm.initializer = create_initializers_from(algo_constants[i])
> onnx.save(model, session.out_path)
> ```
> 
> Note the constants are updated atomically after each evaluation of a graph -- each evaluation always sees the latest set of constants, but they don't change until after the graph has been evaluated.
> 
> As I understand the spec, the result of training should be equivalent to doing this for each iteration:
> 
> 1. Load model
> 2. Run algorithm (equivalent to freezing inference graph as an operator and then running algorithm with GraphCall nodes replaced with that operator)
> 3. Update initializer values from algorithm output based on update_binding
> 4. Save model
> 
> As soon as you've done one iteration the values of the previous iteration are inaccessible (unless you explicitly kept them, or they aren't trainable).
> 
> Does this seem right @wschin?

Yes. The only thing we need to clarify here is `what tensor to use when the user doesn't specify an optional input when calling the inference graph`. In your code, all initializers are used as `global variables`. If I invoke the inference graph twice, the same initializer may be connected to nodes in the two graphs. We can describe such two graph calls in Python:
```python
W_init = Tensor([1, 2, 3])
def inference_graph(X, W=None):
  if W == None:
    W = W_init # W is a reference to W_init
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1)
Y_2 = inference_graph(X_2)
Z = Y_1 + Y_2
```
If we call back-propagation on `Z` agains `W_init`, the trainable tensor `W_init` may receive its gradient `dZ/dW_init`. The user might feel strange because there is no explicit connection between `Z` and `W_init`. Thus, to avoid implicit edges in the training graph, @gramalingam prefers to do this
```
W_init = Tensor([1, 2, 3])
def inference_graph(X, W):
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1, W_init)
Y_2 = inference_graph(X_2, W_init)
Z = Y_1 + Y_2
```
Another way to avoid implicit edges is 
```
W_init = Tensor([1, 2, 3])
def inference_graph(X, W):
  if W == None:
    W = deepcopy(W_init) # W is NOT a reference to W_init
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1, W_init)
Y_2 = inference_graph(X_2, W_init)
Z = Y_1 + Y_2
```

> 
> ### Regarding optional inputs
> It seems the behaviour for optional inputs _within an iteration_ should look identical to the behaviour at inference time. If the input is provided then that tensor name is bound to the input value, if the input is not provided, the tensor name is bound to the constant.

The constant can be passed into optional input slot by a reference or a copy as described in my pseudo code above (by-reference case is the first snippet and the 3rd snippet is for by-copy case).

> 
> I agree that sentence is confusing:
> 
> > Notice that this syntax sugar does NOT eliminate the edge connecting "W" and the GraphCall.
> 
> There isn't really an edge between the `W` tensor and the `GraphCall` node when that input isn't provided -- but there is a default constant `W` which is _in_ the graph called by `GraphCall`, and that is what is used for the tensor name `W` within the graph.

Sorry for repeating my comment. Again, we don't only need to determine which value to use, we also need to determine if the value is pass-by-value or pass-by-reference.

> 
> ### Thoughts
> Some thoughts having done the exercise of writing it out:
> 
> * Should it be specified how data is consumed? Is the flow is `info[0](data[0]); info[1](data[1]); info[0](data[2]); ...`?

No. Input schema of `TrainingInfoProto.algorithm` already encodes such information. If the model author want to train the model iteration-by-iteration, the input shape could be [1, C, H, W]. If the model author thinks a batch should be used, the input shape could be [N, C, H, W]. Similarly, if the author wants to see the entire training data set, we might end up with input shape [10000000, C, H, W].

> * I'm not sure I understand/remember why having multiple training_infos is helpful, unless I misunderstood how they should be used.

It's for GAN. In GAN iterations, two parameters are alternatively updated so there are two assignments in one iteration.
```python
W = W_init
D = D_init
for i in range(10000):
  W = W - f(W, D)
  D = D - f(W, D)
```
To encode two assignments, we need two `TrainingInfoProto`.

> * As I read it, each `algorithm` can only update the inference graph and itself, so each `algorithm` would need to have its own iteration count etc. Is that what you were thinking?

Yes. It allows different stage to determine if its training is successful. In mix-precision training, the user don't need to increase iteration count if gradient is `NaN`.

Reply:
### RE: optional inputs
> we don't only need to determine which value to use, we also need to determine if the value is pass-by-value or pass-by-reference.

It sounds like we agree that one execution of a training_info algorithm looks like:
1. Evaluate the algorithm graph
2. Update the constants _according to update_binding_

Inference (step 1) has no capability for mutation; only the update step (step 2) does, and the update step is precisely defined by the `update_binding` map. The update binding is not affected by whether you use an optional input or not, or what those inputs are -- there is no implicit backprop and update, only explicit binding of which constants to update, and explicit calculation of what the new values should be.

Here's a somewhat contrived example:
```
graph MyInferenceGraph (%x, optional %w) {
  default %w = [1]
  return MatMul(%x, %w)
}
graph algorithm (%data, %label) {
  // We have access to initializers from MyInferenceGraph, namely the default %w

  // Forward pass with no override of %w
  %score = GraphCall(%data, "", graph_name="MyInferenceGraph")

  // Forward pass where we for some reason scale %w by 1.1
  %s = [1.1]
  %larger_w = Mul(%w, %s)
  %score1 = GraphCall(%data, %larger_w, graph_name="MyInferenceGraph")

  // Ensemble the two forward passes
  %ensemble_score = Mean(%score, %score1)

  // Compute delta we want to subtract from %w
  %d_label = Sub(%label, %ensemble_score)
   // dw/dlabel at w=%w (not 1.1*%w)
  %d_w = Gradient(%data, %w, xs=["%data", "%w"], y="%d_label")
  
  // Explicitly calculate new value for %w
  %new_w = Sub(%w, %d_w)
  return %new_w
}
```
and your update_binding would look like:
```python
{ "w": "new_w" }
```

### RE: Thoughts
Thanks for you answers to my questions. Coming back on the data one: so training is only defined for one call; it doesn't loop over a dataset? I.e. the spec defines behaviour of
```python
def train(model, x):
    "Run through training_info once"
    for info in model.training_info:
        outputs = run(info.algorithm, x) # Same input to all infos
        model = update(model, outputs)
    return model
```
and the user can loop over it, or give a generator (e.g. batch generator) to the runtime which will loop over it.

Reply:
> ### RE: optional inputs
> > we don't only need to determine which value to use, we also need to determine if the value is pass-by-value or pass-by-reference.
> 
> It sounds like we agree that one execution of a training_info algorithm looks like:
> 
> 1. Evaluate the algorithm graph
> 2. Update the constants _according to update_binding_
> 
> Inference (step 1) has no capability for mutation; only the update step (step 2) does, and the update step is precisely defined by the `update_binding` map. The update binding is not affected by whether you use an optional input or not, or what those inputs are

Yes!

> -- there is no implicit backprop and update, only explicit binding of which constants to update, and explicit calculation of what the new values should be.

What method is used in your mind for passing optional inputs into the inference graph? Pass-by-value? Pass-by-reference? Pass-by-reference may lead to implicit back propagation edge.

Pass-by-value example:
```python
W_init = Tensor([1, 2, 3]) # Global constant.
def inference_graph(X, W=None): # W is optional.
  if W == None:
    W = deepcopy(W_init) # W is NOT a reference to W_init; we pass W in by value.
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1)
Y_2 = inference_graph(X_2)
Z = Y_1 + Y_2
```

Pass-by-reference example:

```python
W_init = Tensor([1, 2, 3]) # Global constant.
def inference_graph(X, W=None): # W is optional
  if W == None:
    W = W_init # W is a reference to W_init; we pass W in by reference.
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1)
Y_2 = inference_graph(X_2)
Z = Y_1 + Y_2
```

> 
> Here's a somewhat contrived example:
> 
> ```
> graph MyInferenceGraph (%x, optional %w) {
>   default %w = [1]
>   return MatMul(%x, %w)
> }
> graph algorithm (%data, %label) {
>   // We have access to initializers from MyInferenceGraph, namely the default %w
> 
>   // Forward pass with no override of %w
>   %score = GraphCall(%data, "", graph_name="MyInferenceGraph")
> 
>   // Forward pass where we for some reason scale %w by 1.1
>   %s = [1.1]
>   %larger_w = Mul(%w, %s)
>   %score1 = GraphCall(%data, %larger_w, graph_name="MyInferenceGraph")
> 
>   // Ensemble the two forward passes
>   %ensemble_score = Mean(%score, %score1)
> 
>   // Compute delta we want to subtract from %w
>   %d_label = Sub(%label, %ensemble_score)
>    // dw/dlabel at w=%w (not 1.1*%w)
>   %d_w = Gradient(%data, %w, xs=["%data", "%w"], y="%d_label")
>   
>   // Explicitly calculate new value for %w
>   %new_w = Sub(%w, %d_w)
>   return %new_w
> }
> ```

If we choose pass-by-value to pass optional inputs into the first `GraphCall`, the first `GraphCall` has no effect on `%d_w`. In contrast, if we choose pass-by-reference, the following two calls are equivalent

```
%score = GraphCall(%data, "", graph_name="MyInferenceGraph")
```

```
%score = GraphCall(%data, "", graph_name="MyInferenceGraph")
```

Thus, the first `GraphCall` may affect the value of `%d_w`.

It seems to me the first one is more convenient to model authors. The second one is more compiler friendly (because no implicit edges is in computation graph). I think we should do `pass-by-reference` now (yesterday on Gitter, I prefer `pass-by-value`).

> 
> and your update_binding would look like:
> 
> ```python
> { "w": "new_w" }
> ```
> 
> ### RE: Thoughts
> Thanks for you answers to my questions. Coming back on the data one: so training is only defined for one call; it doesn't loop over a dataset? I.e. the spec defines behaviour of
> 
> ```python
> def train(model, x):
>     "Run through training_info once"
>     for info in model.training_info:
>         outputs = run(info.algorithm, x) # Same input to all infos
>         model = update(model, outputs)
>     return model
> ```
> 
> and the user can loop over it, or give a generator (e.g. batch generator) to the runtime which will loop over it.

If the author wants to loop through a data set, the training graph should contain a loop which loops to the given input. The training spec only says `Hey, do this and then your model can be improved` but the end-user really have no idea about the mathematical meaning of the training graph. A training graph can encode 1/1000/any training iterations.

Reply:
@TMVector, in the spec we now clearly say the global initializer `W` would be used in whenever `W` is needed but missing. It's `pass-by-reference` and matches your python code.

Reply:
Short summary of my opinion: specify the W parameter explicitly in the GraphCall (which is the second/middle solution among the three that Wei-Sheng presented). Is there any good reason why we should not do this?

Re. @wschin 's option-1 vs. option-3 : I am not sure I understand the difference here. In fact, I would prefer to simplify both options and write it as
```python
   def inference_graph(X, W=None): # W is optional
      return X * ( W_init if W == None else W)
```
Mathematically, from a semantics-specification, perspective they are the same. The interesting question (which Wei-Sheng refers to) is whether they are equivalent from the perspective of automatic-differentiation. I think they should be. An extra "identity assignment" should not affect or confuse the differentiation-algorithm.

I guess Wei-sheng may be referring to the details of the differentiation-algorithm and the backprop graph generated. Obviously, it could be somewhat more cumbersome depending on the exact option chosen. I feel specifying W explicitly would be simpler than either of the reference/value variant discussed.

Reply:
Let me list the three style again for convenience. I added some comments to Option 3 to explain its difference comparing with Option 1.

Option 1: pass-by-reference (TVMector's python code).

```python
W_init = Tensor([1, 2, 3])
def inference_graph(X, W=None): # W is optional
  if W == None:
    W = W_init # W is a reference to W_init
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1)
Y_2 = inference_graph(X_2)
Z = Y_1 + Y_2
```

Option 2: gramalingam's suggestion. It's the cleanest representation but trainable initializers must be inference inputs.

```python
W_init = Tensor([1, 2, 3])
def inference_graph(X, W): # W is not optional
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1, W_init)
Y_2 = inference_graph(X_2, W_init)
Z = Y_1 + Y_2
```

Option 3: Pass-by-value. Similar to Option 1 but we create an independent copy of `W_init` each time `W`'s value is missing but required. Using Option 3 may generate backward graph different than Option 1.

```python
W_init = Tensor([1, 2, 3])
def inference_graph(X, W=None): # W is optional.
  if W == None:
    # W is NOT a reference to W_init.
    # Calling a copy ctor means that another initializer `W_init_copied` is created and
    # The original `W_init` is not used at all in this scope.
    W = tensor_copy_constructor(W_init) 
  Y = X * W
  return Y
X_1 = Tensor([-1, 0, 1])
X_2 = Tensor([2, 4, 6])
Y_1 = inference_graph(X_1)
Y_2 = inference_graph(X_2)
Z = Y_1 + Y_2
```
I completely agree with @gramalingam that Option 2 is the best to an IR. I prefer Option 1 a little bit more because it doesn't force users to add trainable tensors to inference graph's input list. I might be wrong because I assume that data-scientist users want to directly interact with the IR and I want to make their life easier.

Reply:
I'm so sorry, I completely missed that this was about gradient flow 🤦‍♂.

Yeah okay, I see the merit of passing all trainable tensors as inputs now:
* It would make the behaviour around the `Gradient` operator a lot more obvious
* You need to do this anyway to implement multiple steps of gradient descent within a single `TrainingInfoProto.algorithm` like @wschin mentioned (without just reimplementing the forward graph in the training graph)


 I do still think it's really important that the graph should be able to be optimized for inference without manual effort though, ideally without any changes to the inference graph at all.
* GraphCall's could take constants as inputs too 🤢 maybe via a named-association like @wschin had originally
* Another field on inputs saying whether it should be inference time constant or not :/
* Assume that if the input is also in any `update_binding` then it is inference-time constant (disallows trainable inputs with defaults).

But it raises the question in my mind, can the `Gradient` operator refer to internal tensor names within the inference graph?

However, it seems like the question of gradient flow is bigger than just optional inputs.

If the behaviour of options 1 and 3 are both things we want, what do you think about introducing a `StopGradient` operator, like Pytorch [`torch.Tensor.detach()`](https://pytorch.org/docs/stable/autograd.html#torch.Tensor.detach)?

Reply:
I'm so sorry, I completely missed that this was about gradient flow 🤦‍♂.

Yeah okay, I see the merit of passing all trainable tensors as inputs now, sorry @gramalingam.
* It would make the behaviour around the `Gradient` operator a lot more obvious
* You need to do this anyway to implement multiple steps of gradient descent within a single `TrainingInfoProto.algorithm` like @wschin mentioned (without just reimplementing the forward graph in the training graph)

And @wschin makes a great point that this could change in the future.

I do still think it's really important that the inference graph should be ready for deployment without manual effort though. I'm imagining a use-case where users (possibly without ML expertise) are downloading pre-trained models which they can fine-tune with their own data, but this would also be valuable in other scenarios, just to make deployment easy.
Throwing around some ideas of how that could be achieved:
* `GraphCall` could have the ability to override constants that aren't in the input list. This has the benefit of keeping training and inference semantics quite separate. This could be implemented by listing "new inputs" in an attribute, which would make those named initializers into inputs for this particular `GraphCall`.
  E.g. if the inference graph looked like
  ```
  graph MyInferenceGraph(x) {
    %y = Mul(%x,  %w)
    return %y
  }
  ```
  the `GraphCall` in the training graph might look like
  ```python
  onnx.helper.make_node(
    op_type = "GraphCall",
    inputs = ["x0", "w0"],
    outputs = ["y0"],
    # Attrs
    graph_name = "MyInferenceGraph",
    new_inputs = ["w"],
  )
  ```
* Assume that if an input with a default is also in any `update_binding`, then it is inference-time constant. This disallows trainable, inference-time inputs with defaults.

Side note:
If the behaviour of options 1 and 3 are both things we want, what do you think about introducing a `StopGradient` operator, like Pytorch [`torch.Tensor.detach()`](https://pytorch.org/docs/stable/autograd.html#torch.Tensor.detach)?

Reply:
Thanks @TMVector and @gramalingam for those brilliant ideas.

After reviewing all our discussions, I am shocked for how things could be complicated if we don't use Option 2. What do you guys think if we go with Option 2 and make future steps based on user feedback? In my opinion, it's better to take a more **conservative** step when making such a fundamental architecture decision.

Reply:
Okay, so my understanding is:
* All initializers used within the graph of a `GraphCall`  (e.g. `w`) are treated as though they were detached from gradient flow (i.e. `w' = StopGradient(w)`). This includes initializers that are used when optional inputs aren't provided.
* You can only take the gradient with respect to inputs of `GraphCall`s, but you can pass an initializer as an input to `GraphCall` explicitly. Note only uses of input in the body of the inference graph will pass-through gradient flow, all the uses of the initializer stop the gradient.

Some examples:
```
graph InferenceGraph_1(optional %w = [7]) {
  const c% = [11]
  %y = %w * %c
  return %y
}
graph training() {
  %y1 = GraphCall("", graph_name="InferenceGraph_1")
  // Invalid! %w is not used to calculate %y.
  // We can't calculate any gradients because there are no inputs!
  // The function being differentiated is 'y = StopGradient(w) * StopGradient(c)'.
  // %dy_dw = Gradient(%w, %x0, xs=["%w"], zs=["%x0"], y="%y1")

  // %dy_dw is [11] (StopGradient(c)).
  // The function being differentiated is 'y = w * StopGradient(c)'.
  %y2 = GraphCall(%w, graph_name="InferenceGraph_1")
  %dy_dw = Gradient(%w, xs=["%w"], y="%y2")

  // %dy_dc is [11] (StopGradient(c)), not [22] (2*c).
  // The function being differentiated is 'y = c * StopGradient(c)'.
  %y3 = GraphCall(%c, graph_name="InferenceGraph_1")
  %dy_dc = Gradient(%c, xs=["%c"], y="%y3")

  // %dy_dc is [2*11*11] (2*c*StopGradient(c)'), not [3*11*11] (3*c^2).
  // The function being differentiated is 'y = (c*c) * StopGradient(c)'.
  %c2 = Mul(%c, %c)
  %y3 = GraphCall(%c2, graph_name="InferenceGraph_1")
  %dy_dc = Gradient(%c, xs=["%c"], y="%y3")

  // %dy_dc is [3*11*11] (2*c^2), not [3*11*11 + 7] (3*c^2 + w).
  // The function being differentiated is 'y = c^3 + StopGradient(w) * StopGradient(c)'.
  %c3 = Pow(%c, 3)
  %y4 = GraphCall("", graph_name="InferenceGraph_1")
  %y5 = Add(%c3, %y4)
  %dy_dc = Gradient(%c, xs=["%c"], zs=["%w"], y="%y5")
}
```

Does that sound right?

Reply:
> * All initializers used within the graph of a `GraphCall`  (e.g. `w`) are treated as though they were detached from gradient flow (i.e. `w' = StopGradient(w)`). This includes initializers that are used when optional inputs aren't provided.

Yes before today's WG meeting. Now we stay closer to your original Option 1. Two model level attributes are created for storing global (shared by training and inference) variables:

- ModelProto.global_initializer
- ModelProto.global_sparse_initializer

> * You can only take the gradient with respect to inputs of `GraphCall`s, but you can pass an initializer as an input to `GraphCall` explicitly. Note only uses of input in the body of the inference graph will pass-through gradient flow, all the uses of the initializer stop the gradient.
> 

We can't put any restriction on `Gradient` computation --- it's well-defined by math. It means `You can only take the gradient with respect to inputs of GraphCalls` is not quite valid.